src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2670,5 +2671,260 @@ private interface INegator<T> where T : struct
+                    }
+                    else if (Unsafe.SizeOf<T>() == sizeof(int))
+                    {
+                        // Permute to reverse each vector:
+                        //     +-------------------------------+
+                        //     | A | B | C | D | E | F | G | H |
+                        //     +-------------------------------+
+                        //         --->
+                        //     +-------------------------------+
+                        //     | H | G | F | E | D | C | B | A |
+                        //     +-------------------------------+
+                        tempFirst = Avx2.PermuteVar8x32(tempFirst.As<T, int>(), Vector256.Create(7, 6, 5, 4, 3, 2, 1, 0)).As<int, T>();
+                        tempLast = Avx2.PermuteVar8x32(tempLast.As<T, int>(), Vector256.Create(7, 6, 5, 4, 3, 2, 1, 0)).As<int, T>();
+                    }
+                    else if (Unsafe.SizeOf<T>() == sizeof(long))
+                    {
+                        // Permute to reverse each vector:
+                        tempFirst = Avx2.Permute4x64(tempFirst.As<T, long>(), 0b00_01_10_11).As<long, T>();
+                        tempLast = Avx2.Permute4x64(tempLast.As<T, long>(), 0b00_01_10_11).As<long, T>();
+                    }
+
+                    // Store the values into final location
+                    tempLast.StoreUnsafe(ref elements, firstOffset);
+                    tempFirst.StoreUnsafe(ref elements, lastOffset);
+                }
+                elements = ref Unsafe.Add(ref elements, numIters * numElements);
+                length -= numIters * numElements * 2;
+            }
+            else if (Vector128.IsHardwareAccelerated && (nuint)Vector128<T>.Count * 2 <= length)
+            {
+                nuint numElements = (nuint)Vector128<T>.Count;
+                nuint numIters;
+                // We reverse a span by processing it in the chunks of 16 bytes. We first reverse two 16 byte chunks, one
+                // from the beginning and its corresponding chunk from the end of the span. We then swap these two chunks
+                // chunks to place them at their correct destination. We optimise this further by processing four chunks
+                // of 16 bytes, two from the beginning and two from the end, as much as we can. Then we process two chunks.
+                if ((nuint)Vector128<T>.Count * 4 <= length)
+                {
+                    numIters = (length / numElements) / 4;
+                    for (nuint i = 0; i < numIters; i++)
+                    {
+                        nuint firstOffset = i * numElements * 2;
+                        nuint secondOffset = firstOffset + numElements;
+                        nuint lastOffset = length - (i * numElements * 2) - numElements;
+                        nuint secondLastOffset = lastOffset - numElements;
```

Comment:
Haven't checked impact on performance or code quality on the following patterns
- Reuse the computation, e.g., `nuint lastOffset = length - firstOffset - numElements`
- Extract the computation out of the loop and just increment/decrement offsets by `Vector128<T>.Count`